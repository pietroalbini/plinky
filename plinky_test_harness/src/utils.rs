use anyhow::{bail, Error};
use std::path::{Path, PathBuf};
use std::process::Command;

pub fn run(command: &mut Command) -> Result<(), Error> {
    let cmd_repr = format!("{command:?}");
    let output = command.output()?;
    if !output.status.success() {
        eprintln!("Failed to execute {cmd_repr}");
        eprintln!();
        eprintln!("=== stdout ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stdout));
        eprintln!();
        eprintln!("=== stderr ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stderr));
        eprintln!();
        bail!("command failed with exit {}", output.status);
    }
    Ok(())
}

pub struct RunAndSnapshot {
    name: String,
    path: PathBuf,
    output: String,
}

impl RunAndSnapshot {
    pub fn new(name: &str, path: &Path) -> Self {
        Self { name: name.into(), path: path.into(), output: String::new() }
    }

    pub fn run(&mut self, action: &str, command: &mut Command) -> Result<bool, Error> {
        self.separator();
        match command.output() {
            Ok(output) => {
                self.output.push_str(&format!("{action} exited with {}\n", output.status));
                for (name, content) in [("stdout", &output.stdout), ("stderr", &output.stderr)] {
                    if content.trim_ascii().is_empty() {
                        self.output.push_str(&format!("\nno {name} present\n"));
                    } else {
                        let mut content = String::from_utf8_lossy(content).to_string();
                        content = content.replace(env!("CARGO_MANIFEST_DIR"), "${project}");

                        self.output
                            .push_str(&format!("\n=== {name} ===\n{}\n", content.trim_ascii_end()));
                    }
                }
                Ok(output.status.success())
            }
            Err(err) => {
                self.output.push_str(&format!("{action} failed to execute with error: {err}"));
                Ok(false)
            }
        }
    }

    pub fn note(&mut self, note: &str) {
        self.separator();
        self.output.push_str(note);
        if !note.ends_with('\n') {
            self.output.push('\n');
        }
    }

    fn separator(&mut self) {
        if !self.output.is_empty() {
            self.output.push_str("\n==============\n\n");
        }
    }

    pub fn persist(self) {
        let mut insta_settings = insta::Settings::clone_current();
        insta_settings.set_prepend_module_to_snapshot(false);
        insta_settings.set_omit_expression(true);
        insta_settings.set_snapshot_path(self.path.canonicalize().unwrap());

        insta_settings.bind(|| {
            insta::assert_snapshot!(self.name, self.output.trim_ascii_end());
        });
    }
}

pub(crate) fn file_name(path: impl AsRef<Path>) -> String {
    path.as_ref().file_name().unwrap().to_str().unwrap().to_string()
}

#[track_caller]
pub(crate) fn err_str<T>(result: Result<T, Error>) -> Result<T, String> {
    match result {
        Ok(ok) => Ok(ok),
        Err(err) => {
            let mut repr = format!("error: {err}\n");
            let mut source = err.source();
            while let Some(err) = source {
                repr.push_str(&format!("  cause: {err}\n"));
                source = err.source();
            }
            panic!("{repr}")
        }
    }
}
