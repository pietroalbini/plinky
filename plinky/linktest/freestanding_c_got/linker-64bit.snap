---
source: plinky_test_harness/src/utils.rs
---
linking exited with exit status: 0

no stdout present

=== stderr ===
debug print: loaded object
 │
 │  section .text#1 (perms: rx) in hello.o
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 55 48 89 e5 ba 0c 00 00 00 48 8d 05 00 00 00 00 │ UH.......H...... │
 │   │  │ 48 89 c6 bf 01 00 00 00 ff 15 00 00 00 00 bf 00 │ H............... │
 │   │  │ 00 00 00 ff 15 00 00 00 00                      │ .........        │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   │
 │   │    Relocations:
 │   │  ╭───────────────┬─────────────────────┬────────┬────────────────────╮
 │   │  │ Type          │ Symbol              │ Offset │ Addend             │
 │   │  ├───────────────┼─────────────────────┼────────┼────────────────────┤
 │   │  │ Relative32    │ <section .rodata#5> │ 0xc    │ 0xfffffffffffffffc │
 │   │  ├───────────────┼─────────────────────┼────────┼────────────────────┤
 │   │  │ GOTRelative32 │ write#12            │ 0x1a   │ 0xfffffffffffffffc │
 │   │  ├───────────────┼─────────────────────┼────────┼────────────────────┤
 │   │  │ GOTRelative32 │ exit#13             │ 0x25   │ 0xfffffffffffffffc │
 │   │  ╰───────────────┴─────────────────────┴────────┴────────────────────╯
 │   ┴
 │
 │  section .text#14 (perms: rx) in syscalls.o
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 50 53 51 48 c7 c0 04 00 00 00 48 89 fb 48 89 f1 │ PSQH......H..H.. │
 │   │  │ cd 80 59 5b 58 c3 48 c7 c0 01 00 00 00 48 89 fb │ ..Y[X.H......H.. │
 │   │  │ cd 80                                           │ ..               │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 ┴

debug print: object after relocations are applied
 │
 │  section .got#22 (perms: rw) in <plinky>
 │   │
 │   │  address: 0x402000
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 29 10 40 00 00 00 00 00 3f 10 40 00 00 00 00 00 │ ).@.....?.@..... │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 │
 │  section .text#1 (perms: rx) in hello.o
 │   │
 │   │  address: 0x401000
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 55 48 89 e5 ba 0c 00 00 00 48 8d 05 f0 ef ff ff │ UH.......H...... │
 │   │  │ 48 89 c6 bf 01 00 00 00 ff 15 e2 0f 00 00 bf 00 │ H............... │
 │   │  │ 00 00 00 ff 15 df 0f 00 00                      │ .........        │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 │
 │  section .text#14 (perms: rx) in syscalls.o
 │   │
 │   │  address: 0x401029
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 50 53 51 48 c7 c0 04 00 00 00 48 89 fb 48 89 f1 │ PSQH......H..H.. │
 │   │  │ cd 80 59 5b 58 c3 48 c7 c0 01 00 00 00 48 89 fb │ ..Y[X.H......H.. │
 │   │  │ cd 80                                           │ ..               │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 │
 │    Symbols:
 │  ╭──────────────────────────┬──────────┬────────────┬────────────┬─────────────────╮
 │  │ Name                     │ Type     │ Source     │ Visibility │ Value           │
 │  ├──────────────────────────┼──────────┼────────────┼────────────┼─────────────────┤
 │  │ <symbol#0>               │ none     │ <plinky>   │ local      │ <null>          │
 │  ├──────────────────────────┼──────────┼────────────┼────────────┼─────────────────┤
 │  │ <section .text#1>        │ section  │ hello.o    │ local      │ .text#1 + 0x0   │
 │  ├──────────────────────────┼──────────┼────────────┼────────────┼─────────────────┤
 │  │ <section .rodata#5>      │ section  │ hello.o    │ local      │ .rodata#5 + 0x0 │
 │  ├──────────────────────────┼──────────┼────────────┼────────────┼─────────────────┤
 │  │ _GLOBAL_OFFSET_TABLE_#11 │ none     │ hello.o    │ global     │ <undefined>     │
 │  ├──────────────────────────┼──────────┼────────────┼────────────┼─────────────────┤
 │  │ _start#2                 │ function │ hello.o    │ global     │ .text#1 + 0x0   │
 │  ├──────────────────────────┼──────────┼────────────┼────────────┼─────────────────┤
 │  │ exit#13                  │ none     │ syscalls.o │ global     │ .text#14 + 0x16 │
 │  ├──────────────────────────┼──────────┼────────────┼────────────┼─────────────────┤
 │  │ write#12                 │ none     │ syscalls.o │ global     │ .text#14 + 0x0  │
 │  ╰──────────────────────────┴──────────┴────────────┴────────────┴─────────────────╯
 ┴



